File: Euclid/Game/Room/Room.cs
using Euclid.Messages;
using Euclid.Network;
using Euclid.Network.Streams;
using Euclid.Storage.Database.Data;
using Euclid.Util;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;

namespace Euclid.Game
{
    public class Room
    {
        #region Properties

        public RoomData Data { get; }
        public RoomEntityManager EntityManager { get; }
        public RoomTaskManager TaskManager { get; }
        public RoomItemManager ItemManager { get; }
        public RoomFurniture FurnitureManager { get; set; }
        public RoomMapping Mapping { get; set; }
        public RoomModel Model => RoomManager.Instance.RoomModels.FirstOrDefault(x => x.Data.Id == Data.ModelId);
        public ConcurrentDictionary<IEntity, int> Entities { get; }
        public bool IsActive { get; set; }
        public GameAddress Address;

        #endregion

        #region Constructors

        public Room(RoomData data)
        {
            Data = data;
            Entities = new ConcurrentDictionary<IEntity, int>();
            EntityManager = new RoomEntityManager(this);
            Mapping = new RoomMapping(this);
            TaskManager = new RoomTaskManager(this);
            ItemManager = new RoomItemManager(this);
            FurnitureManager = new RoomFurniture(this);

            if (data.IsPrivateRoom)
                Address = GameServer.Instance.PrivateServer;

            if (data.IsPublicRoom)
                Address = new GameAddress(ServerConfig.Instance.GetString("server", "public", "ip", Data.Id.ToString()), ServerConfig.Instance.GetInt("server", "public", "port", Data.Id.ToString()));
        }

        #endregion

        #region Static methods

        /// <summary>
        /// Wrap the retrieved database data with a room instance >:)
        /// </summary>
        public static Room Wrap(RoomData roomData)
        {
            return new Room(roomData);
        }

        #endregion

        #region Private methods

        /// <summary>
        /// Get if the user has rights
        /// </summary>
        public bool HasRights(int userId, bool checkOwner = true)
        {
            if (checkOwner)
                if (Data.OwnerId == userId)
                    return true;

            var player = PlayerManager.Instance.GetPlayerById(userId);

            if (player != null)
            {
                if (player.UserGroup.HasPermission("room.rights"))
                    return true;
            }

            return false;
        }

        /// <summary>
        /// Get if the user is owner
        /// </summary>
        public bool IsOwner(int userId)
        {
            if (Data.OwnerId == userId)
                return true;

            var player = PlayerManager.Instance.GetPlayerById(userId);

            if (player != null)
            {
                if (player.UserGroup.HasPermission("room.owner"))
                    return true;
            }

            return false;
        }


        /// <summary>
        /// Try and dispose, only if it has 0 players active.
        /// </summary>
        public void TryDispose()
        {
            var playerList = EntityManager.GetEntities<Player>();

            if (playerList.Any())
                return;

            TaskManager.StopTasks();
            RoomManager.Instance.RemoveRoom(Data.Id);

            IsActive = false;
        }

        /// <summary>
        /// Send packet to entire player list in room
        /// </summary>
        public void Send(IMessageComposer composer, List<Player> specificUsers = null)
        {
            if (specificUsers == null)
                specificUsers = EntityManager.GetEntities<Player>();

            foreach (var player in specificUsers)
                player.Send(composer);
        }

        /// <summary>
        /// Forward entity to room
        /// </summary>
        /// <param name="entity"></param>
        public void Forward(IEntity entity)
        {
            if (!(entity is Player))
            {
                return;
            }

            var player = (Player)entity;
            
      
        }

        #endregion
    }
}

File: euclid/src/main/java/euclid/game/room/Room.java
package euclid.game.room;

import euclid.game.Player;
import euclid.game.PlayerManager;
import euclid.game.entity.IEntity;
import euclid.messages.IMessageComposer;
import euclid.network.GameServer;
import euclid.network.streams.GameAddress;
import euclid.util.ServerConfig;
import euclid.storage.database.data.RoomData;

import java.util.List;
import java.util.concurrent.ConcurrentHashMap;

public class Room {

    private final RoomData data;
    private final RoomEntityManager entityManager;
    private final RoomTaskManager taskManager;
    private final RoomItemManager itemManager;
    private RoomFurniture furnitureManager;
    private RoomMapping mapping;
    private final ConcurrentHashMap<IEntity, Integer> entities;
    private boolean isActive;
    private GameAddress address;

    public Room(RoomData data) {
        this.data = data;
        this.entities = new ConcurrentHashMap<>();
        this.entityManager = new RoomEntityManager(this);
        this.mapping = new RoomMapping(this);
        this.taskManager = new RoomTaskManager(this);
        this.itemManager = new RoomItemManager(this);
        this.furnitureManager = new RoomFurniture(this);

        if (data.isPrivateRoom())
            address = GameServer.getInstance().getPrivateServer();

        if (data.isPublicRoom()) {
            String ip = ServerConfig.getInstance().getString("server", "public", "ip", String.valueOf(data.getId()));
            int port = ServerConfig.getInstance().getInt("server", "public", "port", String.valueOf(data.getId()));
            if (ip != null)
                address = new GameAddress(ip, port, data.getId());
        }
    }

    public static Room wrap(RoomData roomData) {
        return new Room(roomData);
    }

    public RoomData getData() {
        return data;
    }

    public RoomEntityManager getEntityManager() {
        return entityManager;
    }

    public RoomTaskManager getTaskManager() {
        return taskManager;
    }

    public RoomItemManager getItemManager() {
        return itemManager;
    }

    public RoomFurniture getFurnitureManager() {
        return furnitureManager;
    }

    public void setFurnitureManager(RoomFurniture furnitureManager) {
        this.furnitureManager = furnitureManager;
    }

    public RoomMapping getMapping() {
        return mapping;
    }

    public void setMapping(RoomMapping mapping) {
        this.mapping = mapping;
    }

    public RoomModel getModel() {
        return RoomManager.getInstance().getRoomModels().stream()
                .filter(x -> x.getData().getId() == data.getModelId())
                .findFirst()
                .orElse(null);
    }

    public ConcurrentHashMap<IEntity, Integer> getEntities() {
        return entities;
    }

    public boolean isActive() {
        return isActive;
    }

    public void setActive(boolean active) {
        isActive = active;
    }

    public GameAddress getAddress() {
        return address;
    }

    public void setAddress(GameAddress address) {
        this.address = address;
    }

    public boolean hasRights(int userId, boolean checkOwner) {
        if (checkOwner && data.getOwnerId() == userId)
            return true;

        Player player = PlayerManager.getInstance().getPlayerById(userId);
        if (player != null) {
            if (player.getUserGroup().hasPermission("room.rights"))
                return true;
        }

        return false;
    }

    public boolean hasRights(int userId) {
        return hasRights(userId, true);
    }

    public boolean isOwner(int userId) {
        if (data.getOwnerId() == userId)
            return true;

        Player player = PlayerManager.getInstance().getPlayerById(userId);
        if (player != null) {
            if (player.getUserGroup().hasPermission("room.owner"))
                return true;
        }

        return false;
    }

    public void tryDispose() {
        List<Player> playerList = entityManager.getEntities(Player.class);

        if (!playerList.isEmpty())
            return;

        taskManager.stopTasks();
        RoomManager.getInstance().removeRoom(data.getId());

        isActive = false;
    }

    public void send(IMessageComposer composer, List<Player> specificUsers) {
        if (specificUsers == null)
            specificUsers = entityManager.getEntities(Player.class);

        for (Player player : specificUsers)
            player.send(composer);
    }

    public void send(IMessageComposer composer) {
        send(composer, null);
    }
}
